src/JsonRpcEngine.ts

```diff
@@ -0,0 +1,341 @@
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { errorCodes, EthereumRpcError, serializeError } from 'eth-rpc-errors';
+
+/**
+ * A String specifying the version of the JSON-RPC protocol.
+ * MUST be exactly "2.0".
+ */
+export type JsonRpcVersion = '2.0';
+
+/**
+ * An identifier established by the Client that MUST contain a String, Number,
+ * or NULL value if included. If it is not included it is assumed to be a
+ * notification. The value SHOULD normally not be Null and Numbers SHOULD
+ * NOT contain fractional parts.
+ */
+export type JsonRpcId = number | string | void;
+
+export interface JsonRpcError {
+  code: number;
+  message: string;
+  data?: unknown;
+  stack?: unknown;
+}
+
+export interface JsonRpcRequest<T> {
+  jsonrpc: JsonRpcVersion;
+  method: string;
+  id: JsonRpcId;
+  params?: T;
+}
+
+export interface JsonRpcNotification<T> {
+  jsonrpc: JsonRpcVersion;
+  method: string;
+  params?: T;
+}
+
+interface JsonRpcResponseBase {
+  jsonrpc: JsonRpcVersion;
+  id: JsonRpcId;
+}
+
+export interface JsonRpcSuccess<T> extends JsonRpcResponseBase {
+  result: T;
+}
+
+export interface JsonRpcFailure extends JsonRpcResponseBase {
+  error: JsonRpcError;
+}
+
+export type JsonRpcResponse<T> = JsonRpcSuccess<T> | JsonRpcFailure;
+
+export type JsonRpcEngineCallbackError = Error | JsonRpcError | null;
+
+export type JsonRpcEngineReturnHandler = (
+  done: (error?: JsonRpcEngineCallbackError) => void
+) => void;
+
+export type JsonRpcEngineNextCallback = (
+  returnHandlerCallback?: JsonRpcEngineReturnHandler
+) => void;
+
+export type JsonRpcEngineEndCallback = (
+  error?: JsonRpcEngineCallbackError
+) => void;
+
+export type JsonRpcMiddleware = (
+  req: JsonRpcRequest<unknown>,
+  res: JsonRpcResponse<unknown>,
+  next: JsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback
+) => void;
+
+interface InternalJsonRpcResponse extends JsonRpcResponseBase {
+  result?: unknown;
+  error?: Error | JsonRpcError;
+  _originalError?: unknown;
+}
+
+type InternalMiddleware = (
+  req: JsonRpcRequest<unknown>,
+  res: InternalJsonRpcResponse,
+  next: JsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback
+) => void;
+
+export class JsonRpcEngine extends SafeEventEmitter {
+  private _middleware: InternalMiddleware[];
+
+  constructor() {
+    super();
+    this._middleware = [];
+  }
+
+  push(middleware: JsonRpcMiddleware): void {
+    this._middleware.push(middleware as InternalMiddleware);
+  }
+
+  handle(
+    request: JsonRpcRequest<unknown>,
+    callback: (
+      error?: JsonRpcEngineCallbackError,
+      response?: JsonRpcResponse<unknown>
+    ) => void
+  ): void;
+
+  handle(
+    requests: JsonRpcRequest<unknown>[],
+    callback: (
+      error?: JsonRpcEngineCallbackError,
+      responses?: JsonRpcResponse<unknown>[]
+    ) => void
+  ): void;
+
+  handle(request: JsonRpcRequest<unknown>): Promise<JsonRpcResponse<unknown>>;
```

Comment:
Ah, that's a neat idea (!), but I don't think we should do it.

The reason being, as a consumer, you will usually want to check what the result is at runtime before doing anything with it. After all, you can't trust the remote to be honest and correct. (I don't even trust _our own background_ to be honest and correct 😅.)

Reply:
hmm... there'd be nothing preventing you from **double checking** that you are aware of the type that should be returned from the remote, but would enforce that you treat the type as you expect it to be... this would all be in userland, there'd be no type enforcement coming from the system -- instead of typing everything as unknown, we'd be pushing the type safety details to the client. They'd be able to then type this in such a way that they couldn't access variables that they themselves do not type as part of the response.

Reply:
Hm, certainly there's nothing preventing consumers from double checking, but it could encourage them to not check at runtime at all. Wouldn't it be better to just leave it as `unknown` and let the user handle the casting after checking it themselves? For example:

```typescript
const response = await engine.handle<undefined, unknown>({
  id: 1,
  jsonrpc: '2.0',
  method: 'foo',
})

if ('result' in response && typeof response.result === 'number') {
  // do stuff
}
```

Reply:
https://www.typescriptlang.org/play?#code/C4TwDgpgBAsghiARhAPAFQHxQLxTVCAD2AgDsATAZygHlEArCAY2CgH4oAFOAJ2AEs4AG3RYAXHigAfKKQCuQodKhyKEAGb9SEcgCgiYAPZ8ooSFABSlQ6Ry6oyxIcNCIcUveXyAtsh6eZSmAeLQBzANkFIQiAbygAbTAeQ0g+EAkgkNJQgF0JKxsAXwiC0nicgG5dXQB6ACo6+zqoAEEoAGVgsKhKSCZ+dRBu4AALaAA3CB5KfhsoQ3VTMct2mgA5AFoAJU4AYSgkw2BDJhcAOibYAFV2-GQCQjgWIRAoACIAJjOABjeLqDqNX0hCMJjM0FKWzATAAalMZnNcAByL7fJFVWoNS4tWz8chkASaKYEIJwRBCfiUMbkKCIV6jaC7CkEpZwVgwG74U6kYBwLRQOAdLrZAA0UDWcl8UxFl2M4quABkFVBxsI5NABlAtEwhHJ8eQzlAAJKLfisSmyI5a0g6vU6LXm6hwSiUSX24606BwS6kI4DfhMNmzUiGtDLVW66DtAASNEVABFLTxvMIXpbWPcJYoBRRxZK-NQY3GFfHLmsaFybLz+eoeE8BDZhAdeMBKP9AcDQaxwZZrKQoUwjTTcD4-MpMt0ZONDHiMQZjOaeVN1E8IX2BwBRHjJHhQGKeU74iSjqZVBzeCAuuChCAZYWhM9QchsuBsCSqADWvoA7qRH6SmA-N8envKpik7BdrRIHgVyYNcbAHLYIAAR3VIJRD3Tx6D7HhoXyddoThaZg0fC9RkMcg7yyB9PDxfCEOhIdHzAXg4G8ShgLQMDqnnEwtGg2D4P7aE1j9TRAwbUgMP3BxsJsXCmHo4TYXhEjPDIkYKKosJmNY9jOO43R+OXVdewYpgkN6GxKAgAAhZ1oBkqA5NIBSlIHIiET-WjKLM5SmN0cDeMXATTMhaF2jkJg4JdDCiBIChqHCizLyMUgbPsmzMIccEJCRV1osvSh0U8HhLwUYAJHgJBUEwQzgqgky4L8gcADE+V1MqHgSqgWuhSy0oyhzstMcBbygJEV34TqIBKhwph3dzoS3Hd6pBSCe2SgbrNqrBcGSyLCtizBlGS9rprkMqMWdEAbSgdRVBYYMoBGdxyFcdAxRhDAAAoytQy9Kr6lKAfQzAAEoJE4ZJvEpVAttSnaUG+rAnO5IIVRwAVvz5Vh1AgYAmBGP6ULQ4AzhchSxRiJxyHSFZ1jOCdsgGEAfriDStKgf6ybOTmaUKcHBdKgnLtscYKb7H7wcC6pjJg0yNxBZgSHIAB1VMCY1wmRhaF0Ce29LHM8Z0bKBpyHAjdVtNFTxinA+XBKgDXFC1tkib1s3uDrdiRusC8bZo8DXooVx2agfm8px13gAAfRxnXPYJpExUpvCJtRFOtV8gBGMUWJ9ygJDif3xqRAu2OKqBCkFs4GVIH7xhwVHaMWRu67GnBsGRAqYqrgAyfvRsgBYVTOMrXSEVhu+RQwGBVpFwZGhwyuAMWoHbieKoFahvYEYQUCVvpVZd1xgG1j39eAQ2bIwcG2DOU2CYfq2IEfYoV9FnhbCRNFApl3QId3qoFPm7ROV9vaVzFEfFWOhQHn3drrK+N8IC-Q5gTTSvkkTRzPvHRBSdgBZzTopDOPws50SgHnZshdi49EMAHCaFd2JImrrXeujdm4jU1O3HsM8Jq9yKovZe3Mv7i3HuVKeD8n7ABfmqCAdsRZr2-hNP+wsgA

--- I updated one of your types to include a string literal so you don't have to check keys to determine the type of response you're dealing with. This is a personal preference, not really related to the question.

If you look at the mock implementation of handle in this playground, and the later invocations, you will see that typescript infers the type of params automatically (no need to cast it to unknown for them, they know what params they are passing and we cannot (yet) enforce them... setting this to unknown doesn't provide any value to the developer but it does provide some value by leaving it as a generic because then they can enforce the types they want in user land. 

Typescript infers the `V` type to be `unknown` by default. If you look at my contrived example, I had to use typecasting to get this to behave correctly which is not functionally different than providing a type to the generic. When we use a generic, however, we get flexibility over further enforcing the structure. I have added a 'Maybe' type here which says "if this is an object, all of the keys could be undefined, otherwise it might be the base type, but it also might be null or undefined...better check everything!"

play with the two example invocations (one using generics, the other one using the default inferred type of unknown)...remove question marks (optional chaining operators) and see the errors you get. Which do you prefer?

Reply:
Alright, you got me! https://github.com/MetaMask/json-rpc-engine/pull/69/commits/e5e294f276ddd973b13dc17b886e48029f7d77e5